src/units.cpp

```diff
@@ -457,43 +459,99 @@ void updateUnitsMap(const UnitsPtr &units, UnitsMap &unitsMap, double exp = 1.0)
                 if (model != nullptr) {
                     auto refUnits = model->units(ref);
                     if ((refUnits == nullptr) || refUnits->isImport()) {
-                        unitsMap.clear();
-                        break;
+                        return false;
+                    }
+                    if (!updateUnitsMap(refUnits, unitsMap, uExp * exp)) {
+                        return false;
                     }
-                    updateUnitsMap(refUnits, unitsMap, uExp * exp);
                 }
             }
         }
     }
+    return true;
 }
 
-UnitsMap createUnitsMap(const UnitsPtr &units)
+UnitsMap createUnitsMap(const UnitsPtr &units, bool &isValid)
 {
     UnitsMap unitsMap;
-    updateUnitsMap(units, unitsMap);
+    isValid = true;
+    if (!updateUnitsMap(units, unitsMap)) {
+        isValid = false;
+        return unitsMap;
+    }
 
     // Checking for exponents of zero in the map, which can be removed.
-    bool requireDimensionless = false;
     auto it = unitsMap.begin();
     while (it != unitsMap.end()) {
         if (it->second == 0.0) {
             it = unitsMap.erase(it);
-            requireDimensionless = true;
-        } else if (it->first == "dimensionless") {
-            it->second = 0.0;
-            ++it;
+        } else if ((it->first == "dimensionless") || (it->first == "radian") || (it->first == "steradian")) {
+            it = unitsMap.erase(it);
         } else {
             ++it;
         }
     }
-    if (requireDimensionless) {
-        unitsMap.emplace(std::make_pair("dimensionless", 0.0));
-    }
-
     return unitsMap;
 }
 
-bool Units::equivalent(const UnitsPtr &units1, const UnitsPtr &units2)
+// UnitsMap createUnitsMap(const UnitsPtr &units)
+// {
+//     UnitsMap unitsMap;
+//     updateUnitsMap(units, unitsMap);
+
+//     // Checking for exponents of zero in the map, which can be removed.
+//     bool requireDimensionless = false;
+//     auto it = unitsMap.begin();
+//     while (it != unitsMap.end()) {
+//         if (it->second == 0.0) {
+//             it = unitsMap.erase(it);
+//             requireDimensionless = true;
+//         }
+//         else if (it->first == "dimensionless") {
+//             it->second = 0.0;
+//             ++it;
+//         }
+//         else {
+//             ++it;
+//         }
+//     }
+
+//     if (requireDimensionless) {
+//         unitsMap.emplace(std::make_pair("dimensionless", 0.0));
+//     }
+
+//     return unitsMap;
+// }
+
+bool Units::usesImportedUnits() const
```

Comment:
Does this function come across as an extension to `isImport`?  It does to me making me think that this belongs as part of the checks to determine if a units is imported or not.

Reply:
It's currently not, no.  But I do think it's better that a unit is called "imported" if it or any of its dependencies are.  Would all our imported things need that kind of check? If so, can move it.

Reply:
We might need to let this one percolate to see if it makes sense tomorrow.

Reply:
... nudge nudge ... @hsorby any more thoughts on this?

Reply:
I have had more thoughts but haven't decided on a course of action.

Reply:
I'm thinking that `isImport` needs to stay as it is but that we might need an additional method like `requiresImports` for components and units that have imported entities as part of their definition.

Reply:
Or maybe look at it from the other side and ask the question `isDefined`.

Reply:
I think that `isDefined` is a little less clear because it lumps together import of self and import of children.  If `isImport` is going to stay as is, then I reckon a `requiresImport` or `usesImport` or `hasChildImport` function would make sense.

Reply:
I've done a bit more thinking ... and I reckon that it doesn't make sense to have a generic function between both `components` and `units` looking for import child dependencies (ie: beyond the `isImport`() function).  Since "children" look different to components and units, there's not really any sense in forcing them to use the same `importedentity` function.  I'm going to change the name of this function to be `requiresImports` and make a separate PR with something similar for the component class.  This PR is all about units and scaling ...